src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -1771,22 +1865,102 @@ public sealed class One : Tests
+                    {
+                        if (TryGetTopLevelLengthTemp(e) is (BoundDagTemp lengthTemp, int offset))
+                        {
+                            switch (test)
+                            {
+                                case BoundDagValueTest t when !t.Value.IsBad:
+                                    {
+                                        Debug.Assert(t.Value.Discriminator == ConstantValueTypeDiscriminator.Int32);
+                                        int lengthValue = safeAdd(t.Value.Int32Value, offset);
+                                        return knownResult(BinaryOperatorKind.Equal, lengthValue) ??
+                                               new One(new BoundDagValueTest(t.Syntax, ConstantValue.Create(lengthValue), lengthTemp));
+                                    }
+                                case BoundDagRelationalTest t when !t.Value.IsBad:
+                                    {
+                                        Debug.Assert(t.Value.Discriminator == ConstantValueTypeDiscriminator.Int32);
+                                        int lengthValue = safeAdd(t.Value.Int32Value, offset);
+                                        return knownResult(t.Relation, lengthValue) ??
+                                               new One(new BoundDagRelationalTest(t.Syntax, t.OperatorKind, ConstantValue.Create(lengthValue), lengthTemp));
+                                    }
+                            }
+                        }
+                        else if (e.Syntax.IsKind(SyntaxKind.ListPattern))
+                        {
+                            switch (test)
+                            {
+                                case BoundDagRelationalTest t when !t.Value.IsBad:
+                                    {
+                                        Debug.Assert(t.Value.Discriminator == ConstantValueTypeDiscriminator.Int32);
+                                        Debug.Assert(t.Relation == BinaryOperatorKind.GreaterThanOrEqual);
+                                        int lengthValue = t.Value.Int32Value;
+                                        return knownResult(t.Relation, lengthValue) ?? this;
+                                    }
+                            }
+                        }
+                    }
+
+                    return this;
+
+                    static Tests? knownResult(BinaryOperatorKind relation, int lengthValue)
+                    {
+                        var lengthValues = ValueSetFactory.ForLength.Related(relation, lengthValue);
+                        if (lengthValues.IsEmpty)
+                            return False.Instance;
+                        if (lengthValues.Complement().IsEmpty)
+                            return True.Instance;
+                        return null;
+                    }
+                    static int safeAdd(int a, int b)
```

Comment:
I'm not sure whether we need `safeAdd`, versus letting bad things happen in unlikely code with max-value list-pattern in source...
Do we have precedent for this sort of thing elsewhere?